Content.Server/Decals/DecalSystem.cs

```diff
@@ -0,0 +1,227 @@
+                    tileIndices.Y != (int) Math.Floor(decal.Coordinates.Y))
+                    continue;
+
+                if (validDelegate == null || validDelegate(decal))
+                {
+                    uids.Add(uid);
+                }
+            }
+
+            return uids;
+        }
+
+        public bool SetDecalPosition(uint uid, EntityCoordinates coordinates)
+        {
+            if (!ChunkIndex.TryGetValue(uid, out var values))
+            {
+                return false;
+            }
+
+            DirtyChunk(values);
+
+            var gridId = coordinates.GetGridId(EntityManager);
+            var decal = ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid];
+            if (gridId == values.gridId)
+            {
+                ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid] =
+                    decal.WithCoordinates(coordinates.Position);
+                return true;
+            }
+
+            ChunkCollections[values.gridId].GetChunk(values.chunkIndices).Remove(uid);
+            var chunkIndices = ChunkCollections[gridId].GetIndices(coordinates.Position);
+            ChunkCollections[gridId].GetChunk(chunkIndices).Add(uid, decal.WithCoordinates(coordinates.Position));
+            ChunkIndex[uid] = (gridId, chunkIndices);
+            DirtyChunk(gridId, chunkIndices);
+            return true;
+        }
+
+        public bool SetDecalColor(uint uid, Color? color)
+        {
+            if (!ChunkIndex.TryGetValue(uid, out var values))
+            {
+                return false;
+            }
+
+            var decal = ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid];
+            ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid] = decal.WithColor(color);
+            DirtyChunk(values);
+            return true;
+        }
+
+        public bool SetDecalId(uint uid, string id)
+        {
+            if (!ChunkIndex.TryGetValue(uid, out var values))
+            {
+                return false;
+            }
+
+            if (!PrototypeManager.HasIndex<DecalPrototype>(id))
+                throw new ArgumentOutOfRangeException($"Tried to set decal id to invalid prototypeid: {id}");
+
+            var decal = ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid];
+            ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid] = decal.WithId(id);
+            DirtyChunk(values);
+            return true;
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            if (_removedDecals.Count > 0)
+                RaiseNetworkEvent(new DecalRemovalUpdateEvent{RemovedDecals = _removedDecals});
+
+            foreach (var playerSession in _playerManager.GetAllPlayers())
+            {
+                if (!_previousSentChunks.TryGetValue(playerSession, out var previous))
+                    previous = new();
```

Comment:
Just listen to the PlayerStatusChange event and then you can also clean these up as people disconnect.